internal/core/app_exec.go

```diff
@@ -0,0 +1,257 @@
+		return err
+	}
+	defer c.Close()
+
+	a.logger.Debug("spooling exec operation")
+
+	_, _, err = a.doOperation(ctx, a.logger.Named("exec"), &execOperation{
+		InstanceId: id,
+		Component:  c,
+		Deployment: d,
+		Client:     a.client,
+	})
+	return err
+}
+
+// execOperation is a simple barebones operation value to conform to the
+// doOperation doOperation interface.
+type execOperation struct {
+	Log        hclog.Logger
+	InstanceId string
+	Component  *Component
+	Deployment *pb.Deployment
+	Client     pb.WaypointClient
+}
+
+func (op *execOperation) Init(app *App) (proto.Message, error) {
+	return op.Deployment, nil
+}
+
+func (op *execOperation) Hooks(app *App) map[string][]*config.Hook {
+	return nil
+}
+
+func (op *execOperation) Labels(app *App) map[string]string {
+	return op.Deployment.Labels
+}
+
+func (op *execOperation) Upsert(
+	ctx context.Context,
+	client pb.WaypointClient,
+	msg proto.Message,
+) (proto.Message, error) {
+	return msg, nil
+}
+
+// pluginExecVirtHandler is an implementation of ceb.VirtualExecHandler
+// that hands off the exec session info to the plugin's Exec function.
+type pluginExecVirtHandler struct {
+	app    *App
+	log    hclog.Logger
+	op     *execOperation
+	execer component.Execer
+	value  *interface{}
+
+	// Set in CreateSession
+	info *ceb.VirtualExecInfo
+
+	// Any window size updates that we get from the virtual CEB
+	wsUpdates chan component.WindowSize
+
+	// wired up to the context running the CEB to allow us the ability
+	// to cancel it from another goroutine
+	cancel func()
+}
+
+// CreateSession just returns itself because we only use one per virtual
+// ceb instance.
+func (p *pluginExecVirtHandler) CreateSession(
+	ctx context.Context,
+	sess *ceb.VirtualExecInfo,
+) (ceb.VirtualExecSession, error) {
+
+	p.log.Info("creating plugin virt handler session")
+
+	p.info = sess
+
+	return p, nil
+}
+
+// Run translates the session info set in CreateSession into the
+// equiv component types and calls the Exec function.
+func (p *pluginExecVirtHandler) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	p.cancel = cancel
+
+	defer func() {
+		p.cancel = nil
+	}()
+
+	esi := &component.ExecSessionInfo{
+		Input:     p.info.Input,
+		Output:    p.info.Output,
+		Error:     p.info.Error,
+		Arguments: p.info.Arguments,
+	}
+
+	if p.info.PTY != nil && p.info.PTY.Enable {
+		p.wsUpdates = make(chan component.WindowSize, 1)
+
+		esi.WindowSizeUpdates = p.wsUpdates
+		esi.IsTTY = true
+		esi.Term = p.info.PTY.Term
+		esi.InitialWindowSize = component.WindowSize{
+			Height: int(p.info.PTY.WindowSize.Height),
+			Width:  int(p.info.PTY.WindowSize.Width),
+		}
+	}
+
+	p.log.Debug("calling plugin with session-info", "arguments", esi.Arguments)
+
+	val, err := p.app.callDynamicFunc(ctx,
+		p.log,
+		nil,
+		p.op.Component,
+		p.execer.ExecFunc(),
+		argNamedAny("deployment", p.op.Deployment.Deployment),
+		argmapper.Named("exec_info", esi),
```

Comment:
Can also do a `argmapper.Typed` here. It doesn't make a big difference but if you don't plan to ever have multiple `ExecSessionInfo` args then its probably less confusing.

Reply:
Ah, gotcha.  👍🏻